plugins/database/mysql/connection_producer.go

```diff
@@ -0,0 +1,223 @@
+	}
+
+	tlsConfig, err := c.getTLSAuth()
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsConfig != nil {
+		if c.tlsConfigName == "" {
+			c.tlsConfigName, err = uuid.GenerateUUID()
+			if err != nil {
+				return nil, fmt.Errorf("unable to generate UUID for TLS configuration: %w", err)
+			}
+		}
+
+		mysql.RegisterTLSConfig(c.tlsConfigName, tlsConfig)
+	}
+
+	// Set initialized to true at this point since all fields are set,
+	// and the connection can be established at a later time.
+	c.Initialized = true
+
+	if verifyConnection {
+		if _, err := c.Connection(ctx); err != nil {
+			return nil, errwrap.Wrapf("error verifying connection: {{err}}", err)
+		}
+
+		if err := c.db.PingContext(ctx); err != nil {
+			return nil, errwrap.Wrapf("error verifying connection: {{err}}", err)
+		}
+	}
+
+	return c.RawConfig, nil
+}
+
+func (c *mySQLConnectionProducer) Connection(ctx context.Context) (interface{}, error) {
+	if !c.Initialized {
+		return nil, connutil.ErrNotInitialized
+	}
+
+	// If we already have a DB, test it and return
+	if c.db != nil {
+		if err := c.db.PingContext(ctx); err == nil {
+			return c.db, nil
+		}
+		// If the ping was unsuccessful, close it and ignore errors as we'll be
+		// reestablishing anyways
+		c.db.Close()
+	}
+
+	connURL, err := c.finalizeConnectionURL()
+
+	c.db, err = sql.Open("mysql", connURL)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set some connection pool settings. We don't need much of this,
+	// since the request rate shouldn't be high.
+	c.db.SetMaxOpenConns(c.MaxOpenConnections)
+	c.db.SetMaxIdleConns(c.MaxIdleConnections)
+	c.db.SetConnMaxLifetime(c.maxConnectionLifetime)
+
+	return c.db, nil
+}
+
+func (c *mySQLConnectionProducer) SecretValues() map[string]interface{} {
+	return map[string]interface{}{
+		c.Password: "[password]",
+	}
+}
+
+// Close attempts to close the connection
+func (c *mySQLConnectionProducer) Close() error {
+	// Grab the write lock
+	c.Lock()
+	defer c.Unlock()
+
+	if c.db != nil {
+		c.db.Close()
+	}
+
+	c.db = nil
+
+	return nil
+}
+
+func (c *mySQLConnectionProducer) getTLSAuth() (tlsConfig *tls.Config, err error) {
+	if len(c.TLSCAData) == 0 &&
+		len(c.TLSCertificateKeyData) == 0 {
+		return nil, nil
+	}
+
+	rootCertPool := x509.NewCertPool()
+	if len(c.TLSCAData) > 0 {
+		ok := rootCertPool.AppendCertsFromPEM(c.TLSCAData)
+		if !ok {
+			return nil, fmt.Errorf("failed to append CA to client options")
+		}
+	}
+
+	clientCert := make([]tls.Certificate, 0, 1)
+
+	if len(c.TLSCertificateKeyData) > 0 {
+		certificate, err := tls.X509KeyPair(c.TLSCertificateKeyData, c.TLSCertificateKeyData)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load tls_certificate_key_data: %w", err)
+		}
+
+		clientCert = append(clientCert, certificate)
+	}
+
+	tlsConfig = &tls.Config{
+		RootCAs:      rootCertPool,
+		Certificates: clientCert,
+	}
+
+	return tlsConfig, nil
+}
+
+func (c *mySQLConnectionProducer) finalizeConnectionURL() (connURL string, err error) {
```

Comment:
What about:
`getDSNWithTLS`
`addTLSToDSN`
`dsnWithTLS`
?

Reply:
Renamed to `addTLStoDSN`. Used lowercase `to` for added readability.